File: dart/lib/src/runtime/source_location.dart
/// Immutable source identities, Unicode-scalar positions, and source spans.
///
/// A [SourceAuthority] owns snapshots of caller-supplied decoded text. Values
/// retain only an opaque authority identity, source identity, scalar offsets,
/// and explicit provenance. They never retain copied source text or parser
/// state.
library;

import 'dart:convert';

const _validateValuePhase = 'validate_value';
const _sourceMismatchCode = 'source_location_source_mismatch';
const _positionOutOfRangeCode = 'source_location_position_out_of_range';
const _reversedSpanCode = 'source_location_reversed_span';
const _invalidDerivedProvenanceCode =
    'source_location_invalid_derived_provenance';

/// Rule and invocation roles attached to private source-location diagnostics.
final class SourceLocationContext {
  const SourceLocationContext({
    required this.ruleRole,
    required this.invocationRole,
  });

  final String ruleRole;
  final String invocationRole;
}

/// The supported materialization policies for explicitly derived text.
enum DerivedTextPolicy {
  concatenateInOrder('concatenate_in_order');

  const DerivedTextPolicy(this.serializedName);

  final String serializedName;
}

/// Immutable source identity plus a zero-based Unicode-scalar offset.
final class Position {
  const Position._({
    required int authorityId,
    required String sourceId,
    required int offset,
  }) : _authorityId = authorityId,
       _sourceId = sourceId,
       _offset = offset;

  final int _authorityId;
  final String _sourceId;
  final int _offset;

  /// The portable zero-based Unicode-scalar offset.
  int get offset => _offset;

  /// Returns a detached neutral record.
  Map<String, Object?> toJson() => <String, Object?>{
    'source_id': _sourceId,
    'offset': _offset,
  };
}

/// Immutable same-source half-open interval plus a provenance label.
final class Span {
  const Span._({
    required int authorityId,
    required String sourceId,
    required int start,
    required int end,
    required String provenance,
  }) : _authorityId = authorityId,
       _sourceId = sourceId,
       _start = start,
       _end = end,
       _provenance = provenance;

  final int _authorityId;
  final String _sourceId;
  final int _start;
  final int _end;
  final String _provenance;

  /// The portable zero-based Unicode-scalar start offset.
  int get start => _start;

  /// The portable number of Unicode scalars in this span.
  int get scalarLength => _end - _start;

  /// Returns a detached neutral record.
  Map<String, Object?> toJson() => _spanRecord(this);
}

/// Immutable ordered direct-span provenance for explicitly derived text.
final class DerivedText {
  DerivedText._({
    required int authorityId,
    required DerivedTextPolicy policy,
    required List<Span> spans,
  }) : _authorityId = authorityId,
       _policy = policy,
       _spans = List<Span>.unmodifiable(spans);

  final int _authorityId;
  final DerivedTextPolicy _policy;
  final List<Span> _spans;

  /// Returns a detached neutral record.
  Map<String, Object?> toJson() => <String, Object?>{
    'policy': _policy.serializedName,
    'spans': <Object?>[for (final span in _spans) span.toJson()],
  };
}

/// Detached line, column, and UTF-8 byte evidence for one position.
final class SourceCoordinates {
  const SourceCoordinates._({
    required String sourceId,
    required int offset,
    required int line,
    required int column,
    required int utf8ByteOffset,
  }) : _sourceId = sourceId,
       _offset = offset,
       _line = line,
       _column = column,
       _utf8ByteOffset = utf8ByteOffset;

  final String _sourceId;
  final int _offset;
  final int _line;
  final int _column;
  final int _utf8ByteOffset;

  /// The one-based line containing this position.
  int get line => _line;

  /// The one-based Unicode-scalar column containing this position.
  int get column => _column;

  /// The zero-based UTF-8 byte offset derived from decoded text.
  int get utf8ByteOffset => _utf8ByteOffset;

  /// Returns a detached neutral record.
  Map<String, Object?> toJson() => <String, Object?>{
    'source_id': _sourceId,
    'offset': _offset,
    'line': _line,
    'column': _column,
    'utf8_byte_offset': _utf8ByteOffset,
  };
}

/// One of the four private immutable-value contract failures.
final class SourceLocationException implements Exception {
  SourceLocationException._({
    required String code,
    required SourceLocationContext context,
    required Map<String, Object?> fields,
  }) : _record = Map<String, Object?>.unmodifiable(<String, Object?>{
         'code': code,
         'phase': _validateValuePhase,
         'rule_role': context.ruleRole,
         'invocation_role': context.invocationRole,
         ...fields,
       });

  final Map<String, Object?> _record;

  /// Returns a detached machine-readable error record.
  Map<String, Object?> toJson() => Map<String, Object?>.from(_record);

  @override
  String toString() => _record['code']! as String;
}

final class _DecodedSource {
  _DecodedSource(this.text) {
    var line = 1;
    var column = 1;
    var codeUnitOffset = 0;
    var utf8ByteOffset = 0;
    final codeUnitAtScalar = <int>[0];
    final lineAtScalar = <int>[line];
    final columnAtScalar = <int>[column];
    final utf8ByteAtScalar = <int>[0];

    for (final scalar in text.runes) {
      final scalarText = String.fromCharCode(scalar);
      codeUnitOffset += scalarText.length;
      utf8ByteOffset += utf8.encode(scalarText).length;
      if (scalar == 0x0a) {
        line += 1;
        column = 1;
      } else {
        column += 1;
      }
      codeUnitAtScalar.add(codeUnitOffset);
      lineAtScalar.add(line);
      columnAtScalar.add(column);
      utf8ByteAtScalar.add(utf8ByteOffset);
    }

    _codeUnitAtScalar = List<int>.unmodifiable(codeUnitAtScalar);
    _lineAtScalar = List<int>.unmodifiable(lineAtScalar);
    _columnAtScalar = List<int>.unmodifiable(columnAtScalar);
    _utf8ByteAtScalar = List<int>.unmodifiable(utf8ByteAtScalar);
  }

  final String text;
  late final List<int> _codeUnitAtScalar;
  late final List<int> _lineAtScalar;
  late final List<int> _columnAtScalar;
  late final List<int> _utf8ByteAtScalar;

  int get scalarLength => _codeUnitAtScalar.length - 1;

  bool containsBoundary(int offset) =>
      offset >= 0 && offset < _codeUnitAtScalar.length;

  int? scalarOffsetAtCodeUnit(int codeUnitOffset) {
    var low = 0;
    var high = _codeUnitAtScalar.length - 1;
    while (low <= high) {
      final middle = low + ((high - low) >> 1);
      final candidate = _codeUnitAtScalar[middle];
      if (candidate == codeUnitOffset) {
        return middle;
      }
      if (candidate < codeUnitOffset) {
        low = middle + 1;
      } else {
        high = middle - 1;
      }
    }
    return null;
  }

  String slice(int start, int end) =>
      text.substring(_codeUnitAtScalar[start], _codeUnitAtScalar[end]);
}

/// Owns immutable decoded-source snapshots and validates typed values.
final class SourceAuthority {
  SourceAuthority({required Map<String, String> sources})
    : _authorityId = _claimAuthorityId(),
      _sources = Map<String, _DecodedSource>.unmodifiable(
        <String, _DecodedSource>{
          for (final entry in sources.entries)
            entry.key: _DecodedSource(entry.value),
        },
      );

  static int _nextAuthorityId = 1;

  final int _authorityId;
  final Map<String, _DecodedSource> _sources;

  static int _claimAuthorityId() => _nextAuthorityId++;

  /// Constructs and validates one Unicode-scalar-offset position.
  Position position({
    required String sourceId,
    required int offset,
    required SourceLocationContext context,
  }) {
    final source = _sources[sourceId];
    if (source == null || !source.containsBoundary(offset)) {
      throw SourceLocationException._(
        code: _positionOutOfRangeCode,
        context: context,
        fields: <String, Object?>{
          'source_id': sourceId,
          'position_offset': offset,
          'source_length': source?.scalarLength ?? 0,
        },
      );
    }
    return Position._(
      authorityId: _authorityId,
      sourceId: sourceId,
      offset: offset,
    );
  }

  /// Converts one current Dart UTF-16 code-unit boundary to a typed position.
  ///
  /// This is runtime-support API. It does not change the engine's existing
  /// code-unit cursor, match, capture, or mark registers.
  Position positionFromCodeUnit({
    required String sourceId,
    required int codeUnitOffset,
    required SourceLocationContext context,
  }) {
    final source = _sources[sourceId];
    final scalarOffset = source?.scalarOffsetAtCodeUnit(codeUnitOffset);
    if (scalarOffset == null) {
      throw SourceLocationException._(
        code: _positionOutOfRangeCode,
        context: context,
        fields: <String, Object?>{
          'source_id': sourceId,
          'position_offset': codeUnitOffset,
          'source_length': source?.scalarLength ?? 0,
        },
      );
    }
    return position(sourceId: sourceId, offset: scalarOffset, context: context);
  }

  /// Returns the Unicode-scalar length of an owned decoded source, if known.
  int? sourceScalarLength(String sourceId) => _sources[sourceId]?.scalarLength;

  /// Constructs and validates one direct half-open span.
  Span directSpan({
    required Position start,
    required Position end,
    required String provenance,
    required SourceLocationContext context,
  }) {
    if (start._sourceId != end._sourceId ||
        start._authorityId != end._authorityId ||
        start._authorityId != _authorityId) {
      throw SourceLocationException._(
        code: _sourceMismatchCode,
        context: context,
        fields: <String, Object?>{
          'source_id': start._sourceId,
          'other_source_id': end._sourceId,
        },
      );
    }
    if (start._offset > end._offset) {
      throw SourceLocationException._(
        code: _reversedSpanCode,
        context: context,
        fields: <String, Object?>{
          'source_id': start._sourceId,
          'start_offset': start._offset,
          'end_offset': end._offset,
        },
      );
    }
    return Span._(
      authorityId: _authorityId,
      sourceId: start._sourceId,
      start: start._offset,
      end: end._offset,
      provenance: provenance,
    );
  }

  /// Constructs and validates explicitly ordered derived text.
  DerivedText derivedText({
    required DerivedTextPolicy policy,
    required List<Span> spans,
    required SourceLocationContext context,
  }) {
    for (var index = 0; index < spans.length; index += 1) {
      final span = spans[index];
      if (span._authorityId != _authorityId ||
          !_sources.containsKey(span._sourceId)) {
        throw SourceLocationException._(
          code: _invalidDerivedProvenanceCode,
          context: context,
          fields: <String, Object?>{
            'provenance_index': index,
            'source_id': span._sourceId,
          },
        );
      }
    }
    return DerivedText._(
      authorityId: _authorityId,
      policy: policy,
      spans: spans,
    );
  }

  /// Derives one-based line/column and UTF-8 byte evidence for a position.
  SourceCoordinates coordinates(
    Position position, {
    required SourceLocationContext context,
  }) {
    final source = _sources[position._sourceId];
    if (position._authorityId != _authorityId ||
        source == null ||
        !source.containsBoundary(position._offset)) {
      throw SourceLocationException._(
        code: _positionOutOfRangeCode,
        context: context,
        fields: <String, Object?>{
          'source_id': position._sourceId,
          'position_offset': position._offset,
          'source_length': source?.scalarLength ?? 0,
        },
      );
    }
    final offset = position._offset;
    return SourceCoordinates._(
      sourceId: position._sourceId,
      offset: offset,
      line: source._lineAtScalar[offset],
      column: source._columnAtScalar[offset],
      utf8ByteOffset: source._utf8ByteAtScalar[offset],
    );
  }

  /// Materializes one direct span or explicitly derived text.
  String materialize(Object value, {required SourceLocationContext context}) {
    if (value is Span) {
      return _materializeSpan(value, context: context, provenanceIndex: 0);
    }
    if (value is DerivedText) {
      if (value._authorityId != _authorityId) {
        throw SourceLocationException._(
          code: _invalidDerivedProvenanceCode,
          context: context,
          fields: <String, Object?>{
            'provenance_index': 0,
            'source_id': value._spans.isEmpty
                ? ''
                : value._spans.first._sourceId,
          },
        );
      }
      final buffer = StringBuffer();
      for (var index = 0; index < value._spans.length; index += 1) {
        buffer.write(
          _materializeSpan(
            value._spans[index],
            context: context,
            provenanceIndex: index,
          ),
        );
      }
      return buffer.toString();
    }
    throw ArgumentError.value(
      value,
      'value',
      'must be a typed direct span or derived text',
    );
  }

  String _materializeSpan(
    Span span, {
    required SourceLocationContext context,
    required int provenanceIndex,
  }) {
    final source = _sources[span._sourceId];
    if (span._authorityId != _authorityId ||
        source == null ||
        !source.containsBoundary(span._start) ||
        !source.containsBoundary(span._end) ||
        span._start > span._end) {
      throw SourceLocationException._(
        code: _invalidDerivedProvenanceCode,
        context: context,
        fields: <String, Object?>{
          'provenance_index': provenanceIndex,
          'source_id': span._sourceId,
        },
      );
    }
    return source.slice(span._start, span._end);
  }
}

Map<String, Object?> _spanRecord(Span span) => <String, Object?>{
  'source_id': span._sourceId,
  'start': span._start,
  'end': span._end,
  'provenance': span._provenance,
};

/// Returns a fresh detached copy of the exact source-boundary projection map.
Map<String, Object?> sourceLocationProjectionRows() => <String, Object?>{
  for (final entry in _sourceLocationProjectionRows.entries)
    entry.key: <Object?>[
      for (final row in entry.value) <Object?>[...row],
    ],
};

/// Returns a fresh detached copy of the exact compatibility alias catalog.
List<Object?> sourceLocationCompatibilityAliases() => <Object?>[
  for (final row in _sourceLocationCompatibilityAliases) <Object?>[...row],
];

const _sourceLocationProjectionRows = <String, List<List<String>>>{
  'capture_mark': <List<String>>[
    ['capture_between', 'span_text'],
    ['capture_from', 'span_text'],
    ['capture_len_between', 'span_length'],
    ['capture_len_from', 'span_length'],
    ['capture_rest', 'span_text'],
    ['capture_rest_from', 'span_text'],
    ['capture_rest_len', 'span_length'],
    ['capture_rest_len_from', 'span_length'],
    ['capture_slice', 'span_text'],
    ['capture_slice_col', 'span_start_column'],
    ['capture_slice_len', 'span_length'],
    ['capture_slice_line', 'span_start_line'],
    ['capture_slice_pos', 'span_start_offset'],
    ['capture_slice_until_cursor', 'span_text'],
    ['capture_slice_until_cursor_len', 'span_length'],
    ['capture_until_boundary', 'span_text'],
    ['capture_take', 'span_text'],
    ['capture_take_between', 'span_text'],
    ['capture_take_between_len', 'span_length'],
    ['capture_take_len', 'span_length'],
    ['capture_take_len_from', 'span_length'],
    ['capture_take_rest', 'span_text'],
    ['capture_take_rest_from', 'span_text'],
    ['capture_take_rest_len', 'span_length'],
    ['capture_take_rest_len_from', 'span_length'],
    ['capture_take_until_cursor', 'span_text'],
    ['capture_take_until_cursor_from', 'span_text'],
    ['capture_take_until_cursor_len', 'span_length'],
    ['capture_take_until_cursor_len_from', 'span_length'],
    ['capture_until_cursor_from', 'span_text'],
    ['capture_until_cursor_len_from', 'span_length'],
    ['mark_capture_slice', 'capture_boundary_write_position'],
    ['mark_copy', 'mark_write_position'],
    ['mark_exists', 'mark_exists'],
    ['mark_here', 'mark_write_position'],
    ['mark_input_end', 'mark_write_position'],
    ['mark_input_start', 'mark_write_position'],
    ['mark_pos', 'mark_read_offset'],
    ['start_capture_slice', 'capture_boundary_write_position'],
    ['start_capture_slice_from', 'capture_boundary_write_position'],
    ['clear_mark', 'mark_delete'],
    ['mark_col', 'mark_read_column'],
    ['mark_entry_end', 'mark_write_position'],
    ['mark_entry_start', 'mark_write_position'],
    ['mark_line', 'mark_read_line'],
    ['mark_match_end', 'mark_write_position'],
    ['mark_match_start', 'mark_write_position'],
  ],
  'entry_match': <List<String>>[
    ['entry_col', 'span_start_column'],
    ['entry_end_col', 'position_column'],
    ['entry_end_line', 'position_line'],
    ['entry_end_pos', 'position_offset'],
    ['entry_group', 'capture_group_text'],
    ['entry_groups', 'capture_group_list'],
    ['entry_has', 'capture_group_exists'],
    ['entry_len', 'span_length'],
    ['entry_line', 'span_start_line'],
    ['entry_map', 'capture_group_map'],
    ['entry_named', 'capture_group_text'],
    ['entry_start_col', 'span_start_column'],
    ['entry_start_line', 'span_start_line'],
    ['entry_start_pos', 'span_start_offset'],
    ['entry_text', 'span_text'],
    ['match_col', 'span_start_column'],
    ['match_end_col', 'position_column'],
    ['match_end_line', 'position_line'],
    ['match_end_pos', 'position_offset'],
    ['match_group', 'capture_group_text'],
    ['match_groups', 'capture_group_list'],
    ['match_has', 'capture_group_exists'],
    ['match_len', 'span_length'],
    ['match_line', 'span_start_line'],
    ['match_map', 'capture_group_map'],
    ['match_named', 'capture_group_text'],
    ['match_start_col', 'span_start_column'],
    ['match_start_line', 'span_start_line'],
    ['match_start_pos', 'span_start_offset'],
    ['match_text', 'span_text'],
  ],
  'input_cursor': <List<String>>[
    ['cursor_col', 'position_column'],
    ['cursor_line', 'position_line'],
    ['cursor_pos', 'cursor_position'],
    ['cursor_rest', 'span_text'],
    ['cursor_rest_len', 'span_length'],
    ['input_end_col', 'position_column'],
    ['input_end_line', 'position_line'],
    ['input_end_pos', 'position_offset'],
    ['input_len', 'source_length'],
    ['input_slice', 'source_slice_text'],
    ['input_text', 'source_text'],
  ],
  'cursor_control': <List<String>>[
    ['restore_cursor', 'cursor_state_write_compatibility'],
    ['rewind_entry_start', 'cursor_state_write_compatibility'],
    ['rewind_match_start', 'cursor_state_write_compatibility'],
    ['save_cursor', 'cursor_checkpoint_compatibility'],
  ],
};

const _sourceLocationCompatibilityAliases = <List<String>>[
  ['capture_from_rule_start', 'capture_slice'],
  ['capture_len_from_rule_start', 'capture_slice_len'],
  ['capture_rest_length', 'capture_rest_len'],
  ['capture_slice_here', 'start_capture_slice'],
  ['capture_slice_length', 'capture_slice_len'],
  ['entry_named_map', 'entry_map'],
  ['match_named_map', 'match_map'],
];

File: julia/src/runtime/SourceLocation.jl
module SourceLocation

const _VALIDATE_VALUE_PHASE = "validate_value"
const _SOURCE_MISMATCH_CODE = "source_location_source_mismatch"
const _POSITION_OUT_OF_RANGE_CODE = "source_location_position_out_of_range"
const _REVERSED_SPAN_CODE = "source_location_reversed_span"
const _INVALID_DERIVED_PROVENANCE_CODE =
    "source_location_invalid_derived_provenance"
const _NEXT_AUTHORITY_ID = Base.Threads.Atomic{UInt64}(0)

"""Rule and invocation roles attached to private source-location diagnostics."""
struct SourceLocationContext
    rule_role::String
    invocation_role::String

    function SourceLocationContext(
        ::Val{:internal},
        rule_role::String,
        invocation_role::String,
    )
        return new(rule_role, invocation_role)
    end
end

function SourceLocationContext(;
    rule_role::AbstractString,
    invocation_role::AbstractString,
)
    return SourceLocationContext(
        Val(:internal),
        String(rule_role),
        String(invocation_role),
    )
end

"""Materialization policies for explicitly derived text."""
@enum DerivedTextPolicy begin
    ConcatenateInOrder
end

"""Immutable source identity plus a zero-based Unicode-scalar offset."""
struct Position
    _authority_id::UInt64
    _source_id::String
    _offset::Int

    function Position(
        ::Val{:internal},
        authority_id::UInt64,
        source_id::String,
        offset::Int,
    )
        return new(authority_id, source_id, offset)
    end
end

"""Immutable same-source half-open interval plus explicit provenance."""
struct Span
    _authority_id::UInt64
    _source_id::String
    _start::Int
    _stop::Int
    _provenance::String

    function Span(
        ::Val{:internal},
        authority_id::UInt64,
        source_id::String,
        start::Int,
        stop::Int,
        provenance::String,
    )
        return new(authority_id, source_id, start, stop, provenance)
    end
end

"""Immutable ordered direct-span provenance for explicitly derived text."""
struct DerivedText
    _authority_id::UInt64
    _policy::DerivedTextPolicy
    _spans::Tuple{Vararg{Span}}

    function DerivedText(
        ::Val{:internal},
        authority_id::UInt64,
        policy::DerivedTextPolicy,
        spans::Tuple{Vararg{Span}},
    )
        return new(authority_id, policy, spans)
    end
end

"""Detached line, column, and UTF-8 byte evidence for one position."""
struct SourceCoordinates
    _source_id::String
    _offset::Int
    _line::Int
    _column::Int
    _utf8_byte_offset::Int

    function SourceCoordinates(
        ::Val{:internal},
        source_id::String,
        offset::Int,
        line::Int,
        column::Int,
        utf8_byte_offset::Int,
    )
        return new(source_id, offset, line, column, utf8_byte_offset)
    end
end

"""One of the four private immutable-value contract failures."""
struct SourceLocationException <: Exception
    _record::Tuple{Vararg{Pair{String,Any}}}

    function SourceLocationException(
        ::Val{:internal},
        record::Tuple{Vararg{Pair{String,Any}}},
    )
        return new(record)
    end
end

Base.showerror(io::IO, error::SourceLocationException) =
    print(io, first(error._record).second)

struct _DecodedSource
    source_id::String
    text::String
    codeunit_at_scalar::Tuple{Vararg{Int}}
    line_at_scalar::Tuple{Vararg{Int}}
    column_at_scalar::Tuple{Vararg{Int}}
    utf8_byte_at_scalar::Tuple{Vararg{Int}}
end

"""Authority that owns immutable decoded-source snapshots and validates values."""
struct SourceAuthority
    _authority_id::UInt64
    _sources::Tuple{Vararg{_DecodedSource}}

    function SourceAuthority(
        ::Val{:internal},
        authority_id::UInt64,
        sources::Tuple{Vararg{_DecodedSource}},
    )
        return new(authority_id, sources)
    end
end

function SourceAuthority(;
    sources::AbstractDict{<:AbstractString,<:AbstractString},
)
    owned_sources = _DecodedSource[
        _decoded_source(source_id, decoded_text)
        for (source_id, decoded_text) in pairs(sources)
    ]
    sort!(owned_sources; by = source -> source.source_id)
    return SourceAuthority(
        Val(:internal),
        _claim_authority_id(),
        Tuple(owned_sources),
    )
end

function _claim_authority_id()
    while true
        previous = _NEXT_AUTHORITY_ID[]
        previous == typemax(UInt64) && error("source authority identity space exhausted")
        next_id = previous + UInt64(1)
        Base.Threads.atomic_cas!(_NEXT_AUTHORITY_ID, previous, next_id) == previous &&
            return next_id
    end
end

function _decoded_source(source_id::AbstractString, decoded_text::AbstractString)
    owned_text = String(Vector{UInt8}(codeunits(String(decoded_text))))
    codeunit_at_scalar = Int[0]
    line_at_scalar = Int[1]
    column_at_scalar = Int[1]
    utf8_byte_at_scalar = Int[0]
    codeunit_offset = 0
    utf8_byte_offset = 0
    line = 1
    column = 1

    for scalar in owned_text
        width = ncodeunits(string(scalar))
        codeunit_offset += width
        utf8_byte_offset += width
        if scalar == '\n'
            line += 1
            column = 1
        else
            column += 1
        end
        push!(codeunit_at_scalar, codeunit_offset)
        push!(line_at_scalar, line)
        push!(column_at_scalar, column)
        push!(utf8_byte_at_scalar, utf8_byte_offset)
    end

    return _DecodedSource(
        String(source_id),
        owned_text,
        Tuple(codeunit_at_scalar),
        Tuple(line_at_scalar),
        Tuple(column_at_scalar),
        Tuple(utf8_byte_at_scalar),
    )
end

function _source(authority::SourceAuthority, source_id::AbstractString)
    source_id_string = String(source_id)
    for source in authority._sources
        source.source_id == source_id_string && return source
    end
    return nothing
end

_scalar_length(source::_DecodedSource) = length(source.codeunit_at_scalar) - 1

function _boundary_index(source::_DecodedSource, offset::Int)
    0 <= offset <= _scalar_length(source) || return nothing
    return offset + 1
end

function _throw_value_error(
    code::AbstractString,
    context::SourceLocationContext,
    fields::Pair...,
)
    record = Pair{String,Any}[
        "code" => String(code),
        "phase" => _VALIDATE_VALUE_PHASE,
        "rule_role" => context.rule_role,
        "invocation_role" => context.invocation_role,
    ]
    for field in fields
        push!(record, String(field.first) => field.second)
    end
    throw(SourceLocationException(Val(:internal), Tuple(record)))
end

"""Construct and validate one Unicode-scalar-offset position."""
function position(
    authority::SourceAuthority;
    source_id::AbstractString,
    offset::Int,
    context::SourceLocationContext,
)
    source_id_string = String(source_id)
    source = _source(authority, source_id_string)
    if source === nothing || _boundary_index(source, offset) === nothing
        _throw_value_error(
            _POSITION_OUT_OF_RANGE_CODE,
            context,
            "source_id" => source_id_string,
            "position_offset" => offset,
            "source_length" => (source === nothing ? 0 : _scalar_length(source)),
        )
    end
    return Position(
        Val(:internal),
        authority._authority_id,
        source_id_string,
        offset,
    )
end

"""Construct and validate one direct half-open span."""
function direct_span(
    authority::SourceAuthority;
    start::Position,
    stop::Position,
    provenance::AbstractString,
    context::SourceLocationContext,
)
    if start._source_id != stop._source_id ||
            start._authority_id != stop._authority_id ||
            start._authority_id != authority._authority_id
        _throw_value_error(
            _SOURCE_MISMATCH_CODE,
            context,
            "source_id" => start._source_id,
            "other_source_id" => stop._source_id,
        )
    end
    if start._offset > stop._offset
        _throw_value_error(
            _REVERSED_SPAN_CODE,
            context,
            "source_id" => start._source_id,
            "start_offset" => start._offset,
            "end_offset" => stop._offset,
        )
    end
    provenance_string = String(provenance)
    isempty(provenance_string) && throw(ArgumentError("span provenance must not be empty"))
    return Span(
        Val(:internal),
        authority._authority_id,
        start._source_id,
        start._offset,
        stop._offset,
        provenance_string,
    )
end

"""Construct and validate explicitly ordered derived text."""
function derived_text(
    authority::SourceAuthority;
    policy::DerivedTextPolicy,
    spans::AbstractVector,
    context::SourceLocationContext,
)
    policy == ConcatenateInOrder || throw(ArgumentError("unsupported derived-text policy"))
    owned_spans = Span[]
    for (one_based_index, value) in enumerate(spans)
        value isa Span || throw(ArgumentError("derived-text spans must contain only Span values"))
        if value._authority_id != authority._authority_id ||
                _source(authority, value._source_id) === nothing
            _throw_value_error(
                _INVALID_DERIVED_PROVENANCE_CODE,
                context,
                "provenance_index" => one_based_index - 1,
                "source_id" => value._source_id,
            )
        end
        push!(owned_spans, value)
    end
    return DerivedText(
        Val(:internal),
        authority._authority_id,
        policy,
        Tuple(owned_spans),
    )
end

"""Derive one-based line/column and UTF-8 byte evidence for a position."""
function coordinates(
    authority::SourceAuthority,
    value::Position;
    context::SourceLocationContext,
)
    source = _source(authority, value._source_id)
    index = source === nothing ? nothing : _boundary_index(source, value._offset)
    if value._authority_id != authority._authority_id || index === nothing
        _throw_value_error(
            _POSITION_OUT_OF_RANGE_CODE,
            context,
            "source_id" => value._source_id,
            "position_offset" => value._offset,
            "source_length" => (source === nothing ? 0 : _scalar_length(source)),
        )
    end
    return SourceCoordinates(
        Val(:internal),
        value._source_id,
        value._offset,
        source.line_at_scalar[index],
        source.column_at_scalar[index],
        source.utf8_byte_at_scalar[index],
    )
end

function _invalid_provenance(
    context::SourceLocationContext,
    provenance_index::Int,
    source_id::AbstractString,
)
    _throw_value_error(
        _INVALID_DERIVED_PROVENANCE_CODE,
        context,
        "provenance_index" => provenance_index,
        "source_id" => String(source_id),
    )
end

function _materialize_span(
    authority::SourceAuthority,
    value::Span,
    context::SourceLocationContext,
    provenance_index::Int,
)
    source = _source(authority, value._source_id)
    start_index = source === nothing ? nothing : _boundary_index(source, value._start)
    stop_index = source === nothing ? nothing : _boundary_index(source, value._stop)
    if value._authority_id != authority._authority_id ||
            source === nothing ||
            start_index === nothing ||
            stop_index === nothing ||
            value._start > value._stop
        _invalid_provenance(context, provenance_index, value._source_id)
    end
    value._start == value._stop && return ""
    first_codeunit = source.codeunit_at_scalar[start_index] + 1
    after_last_codeunit = source.codeunit_at_scalar[stop_index] + 1
    last_codeunit = prevind(source.text, after_last_codeunit)
    return String(SubString(source.text, first_codeunit, last_codeunit))
end

"""Materialize one direct span or explicitly derived text value."""
function materialize(
    authority::SourceAuthority,
    value::Span;
    context::SourceLocationContext,
)
    return _materialize_span(authority, value, context, 0)
end

function materialize(
    authority::SourceAuthority,
    value::DerivedText;
    context::SourceLocationContext,
)
    if value._authority_id != authority._authority_id
        source_id = isempty(value._spans) ? "" : first(value._spans)._source_id
        _invalid_provenance(context, 0, source_id)
    end
    return join(
        _materialize_span(authority, span, context, index - 1)
        for (index, span) in enumerate(value._spans)
    )
end

"""Return a fresh detached neutral record for a typed source value."""
function to_json(value::Position)
    return Dict{String,Any}(
        "source_id" => value._source_id,
        "offset" => value._offset,
    )
end

function to_json(value::Span)
    return Dict{String,Any}(
        "source_id" => value._source_id,
        "start" => value._start,
        "end" => value._stop,
        "provenance" => value._provenance,
    )
end

function to_json(value::DerivedText)
    return Dict{String,Any}(
        "policy" => "concatenate_in_order",
        "spans" => Any[to_json(span) for span in value._spans],
    )
end

function to_json(value::SourceCoordinates)
    return Dict{String,Any}(
        "source_id" => value._source_id,
        "offset" => value._offset,
        "line" => value._line,
        "column" => value._column,
        "utf8_byte_offset" => value._utf8_byte_offset,
    )
end

function to_json(error::SourceLocationException)
    return Dict{String,Any}(field.first => field.second for field in error._record)
end

end
